go/vt/vttablet/tabletserver/tx_engine.go

```diff
@@ -149,38 +145,74 @@ func (te *TxEngine) transition(state txEngineState) {
 	}
 
 	log.Infof("TxEngine transition: %v", state)
-	switch te.state {
-	case AcceptingReadOnly, AcceptingReadAndWrite:
+
+	// When we are transitioning from read write state, we should close all transactions.
+	if te.state == AcceptingReadAndWrite {
 		te.shutdownLocked()
-	case NotServing:
-		// No special action.
 	}
 
 	te.state = state
+	if te.twopcEnabled && te.state == AcceptingReadAndWrite {
+		// If the prepared pool is not open, then we need to redo the prepared transactions
+		// before we open the transaction engine to accept new writes.
+		// This check is required because during a Promotion, we would have already setup the prepared pool
+		// and redid the prepared transactions when we turn super_read_only off. So we don't need to do it again.
+		if !te.preparedPool.IsOpen() {
+			// We need to redo prepared transactions here to handle vttablet restarts.
+			// If MySQL continues to work fine, then we won't end up redoing the prepared transactions as part of any RPC call
+			// since VTOrc won't call `UndoDemotePrimary`. We need to do them as part of this transition.
+			_ = te.redoPreparedTransactionsLocked()
```

Comment:
I think we should prevent doing shutdown of tx engine twice

Reply:
We aren't doing it twice. `te.state` at the start of the function and `state` in the function call have to be different, otherwise we return from the function much sooner.

So, the first shutdown runs when `te.state` is `AcceptingReadWrite` and the second one only runs when `state` is `AcceptingReadWrite`

Reply:
agreed.